base_layer/wallet/src/storage/sqlite_db/wallet.rs

```diff
@@ -104,6 +104,82 @@ impl Argon2Parameters {
     }
 }
 
+/// A structure to hold key-related database field data, to make atomic operations cleaner
+pub struct DatabaseKeyFields {
+    secondary_key_version: u8,   // the encryption parameter version
+    secondary_key_salt: String,  // the high-entropy salt used to derive the secondary key
+    encrypted_main_key: Vec<u8>, // the main key, encrypted with the secondary key
+}
+impl DatabaseKeyFields {
+    /// Read and parse field data from the database atomically
+    pub fn read(connection: &SqliteConnection) -> Result<Option<Self>, WalletStorageError> {
+        let mut secondary_key_version: Option<String> = None;
+        let mut secondary_key_salt: Option<String> = None;
+        let mut encrypted_main_key: Option<String> = None;
+
+        // Read all fields atomically
+        connection
+            .transaction::<_, Error, _>(|| {
+                secondary_key_version = WalletSettingSql::get(&DbKey::SecondaryKeyVersion, connection)
+                    .map_err(|_| Error::RollbackTransaction)?;
```

Comment:
This mapping is not actually needed. Any Err will rollback the transaction. The docs on diesel.rs don't reflect this in v1.4, but have been updated in the 2.0 docs on docs.rs. In either case, I checked the source and it rolls back on Err.

So it's better to just return the error with `?` and have the original error find their way to the logs

Reply:
I had tried this initially, but the compiler complained that the internal error couldn't be converted to a `diesel` error.